src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -9088,6 +9088,30 @@ bool MergeTreeData::initializeDiskOnConfigChange(const std::set<String> & new_ad
     return true;
 }
 
+void MergeTreeData::loadPrimaryKeys()
+{
+    /// Thread pool to process parts within each table in parallel
+    auto & thread_pool = DB::getActivePartsLoadingThreadPool().get();
+
+    DataPartStates affordable_states = { MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated, MergeTreeDataPartState::Deleting };
+    for (const auto & data_part : getDataParts(affordable_states))
+    {
+        if (data_part->isProjectionPart())
+            continue;
+
+        if (!data_part->isIndexLoaded())
+        {
+            /// Use thread pool to parallelize part loading
+            thread_pool.scheduleOrThrowOnError([data_part] {
+                const_cast<IMergeTreeDataPart &>(*data_part).getIndex();
+            });
+        }
+    }
+
+    /// Wait for all parts to finish loading
+    thread_pool.wait();
```

Comment:
Afaik, `wait` will wait for all active threads in `getActivePartsLoadingThreadPool` (l. 9094). This can include threads which are unrelated to index loading.

In my view, we should keep this simple for now (as simple as `unloadPrimaryKeys`, l. 9115) and not use threading at all. After all, manual index loading is a rare special case which isn't necessary in practice unless for benchmarking etc.

Reply:
Alright, I'll keep it single-threaded. If we ever want to use multiple threads, we should use `ThreadPoolCallbackRunnerLocal` which accepts a pool. That way, we can wait only for those tasks we care about.